Update Shell completion

jakecastelli opened the pull request:
Fixes: #918 

The list of options [[ref](https://quick-lint-js.com/cli/#_options)]:

`*` means (helper text may need to change)

**zsh:**
- [x] --output-format=format *
- [x] --diagnostic-hyperlinks=when
- [x] --vim-file-bufnr=number *
- [x] --path-for-config-search=path (new)
- [x] --config-file=file
- [x] --language=languageid (new)
- [x] --exit-fail-on=errors *
- [x] --stdin
- [x] --lsp / --lsp-server
- [x] --snarky
- [x] -h / --help *
- [x] --debug-apps (new)
- [x] -v / --version *

**bash:**
- [x] --output-format=format
- [x] --diagnostic-hyperlinks=when
- [x] --vim-file-bufnr=number
- [x] --path-for-config-search=path (new)
- [x] --config-file=file
- [x] --language=languageid (new)
- [x] --exit-fail-on=errors
- [x] --stdin
- [x] --lsp / --lsp-server
- [x] --snarky (new)
- [x] -h / --help
- [x] --debug-apps (new)
- [x] -v / --version

**fish:**
- [x] --output-format=format
- [x] --diagnostic-hyperlinks=when
- [x] --vim-file-bufnr=number
- [x] --path-for-config-search=path (new)
- [x] --config-file=file
- [x] --language=languageid (new)
- [x] --exit-fail-on=errors
- [x] --stdin
- [x] --lsp / --lsp-server
- [x] --snarky
- [x] -h / --help
- [x] --debug-apps (new)
- [x] -v / --version

powershell:
- [x] --output-format=format *
- [x] --diagnostic-hyperlinks=when
- [x] --vim-file-bufnr=number
- [x] --path-for-config-search=path (new)
- [x] --config-file=file
- [x] --language=languageid (new)
- [x] --exit-fail-on=errors
- [x] --stdin
- [x] --lsp / --lsp-server
- [x] --snarky
- [x] -h / --help *
- [x] --debug-apps (new)
- [x] -v / --version *

jakecastelli commented:
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

strager commented on completions/quick-lint-js.bash:
```diff
@@ -14,14 +18,18 @@ _quick-lint-js () {
                         ;;
                 --output-format=*)
-                        COMPREPLY=($(compgen -W 'gnu-like vim-qflist-json emacs-lisp' -- "${cur#*=}"))
+                        COMPREPLY=($(compgen -W '${_quick_lint_js_output_formats}' -- "${cur#*=}"))
```
Doesn't `'` (single quote) disable variable expansion? Don't we want `"${_quick_lint_js_output_formats}"` (double quotes) instead?

jakecastelli replied:
<img width="855" alt="Screenshot 2023-04-18 at 11 21 49 am" src="https://user-images.githubusercontent.com/38635403/232649339-3a596394-ddff-4c0c-9cc0-ed2e0e432aa3.png">

single quote seems like working on bash 5.2, but I think double quote would cause less confusion because I believe single quote wasn't working in the past.
I will change them to double quote 👍

strager commented:
> For bash completion, I am using v5.2 and I have to update to `bash-completion@2` in order to make it work. Not sure it was just me or something we need to update to README docs.

Hmm, I didn't know that. It's definitely not your fault, of course.

strager reviewed (approved):
Overall, this looks good to me. :+1: But see my inline comment.

jakecastelli commented:
Before this land, I would like to test bash completion on Linux to verify if the doc is outdated, if it is I will update the README doc for bash.

- [ ] verify bash completion instruction

jakecastelli commented:
Hi @strager does `quick-lint-js` has a docker image? I want to test the changes in a docker container which seems more easily and less interference.

strager commented:
> Hi @strager does `quick-lint-js` has a docker image?

I don't know what you mean. What would the Docker image contain?

jakecastelli commented:
> What would the Docker image contain?
As far as I am thinking the docker image could contain the `quick-lint-js` binary or an environment that could build `quick-lint-js` from source.

The container could act as a sandbox environment for people who would like to try it out locally, I have seen some neovim distro does it, they provide you instructions to have a play with their distro in a docker container before you decide if you wanna install it.

strager commented:
> the docker image could contain the `quick-lint-js` binary or an environment that could build `quick-lint-js` from source.

CI uses the `ghcr.io/quick-lint/quick-lint-js-github-builder` Docker images. Dockerfile is in [`.github/docker/quick-lint-js-github-builder/`](https://github.com/quick-lint/quick-lint-js/tree/f05ef2ef06d03a053a890797a61e4922dfe42a55/.github/docker/quick-lint-js-github-builder).

> The container could act as a sandbox environment for people who would like to try it out locally, I have seen some neovim distro does it, they provide you instructions to have a play with their distro in a docker container before you decide if you wanna install it.

That's a good idea, actually.

strager commented on completions/quick-lint-js.bash:
```diff
@@ -14,20 +18,24 @@ _quick-lint-js () {
+                        COMPREPLY=($(compgen -W "${_quick_lint_js_diagnostic_hyperlink_modes}" -- "${cur#*=}"))
+                        return
+                        ;;
+                --exit-fail-on=*|--vim-file-bufnr=*|--path-for-config-search=*)
```
`--path-for-config-search` accepts a (possibly-fake) file path, but pressing tab after typing `--path-for-config-search=` offers no completions.

This was broken before your patch too, but `--path-for-config-search` wasn't mentioned before your patch.

strager commented on completions/quick-lint-js.bash:
```diff
@@ -1,11 +1,15 @@
+_quick_lint_js_languages="default javascript javascript-jsx experimental-typescript experimental-typescript-jsx"
+
 _quick-lint-js () {
         local cur opts
         _init_completion -n = || return
 
-        opts='--help --version --lsp-server --config-file= --stdin --exit-fail-on= --output-format= --diagnostic-hyperlinks= --vim-file-bufnr='
+        opts="--help --version --lsp --lsp-server --config-file= --stdin --exit-fail-on= --output-format= --diagnostic-hyperlinks= --vim-file-bufnr= --language= --path-for-config-search= --snarky --debug-apps"
```
`--lsp` and `--lsp-server` are the same. (`--lsp` exists only for people who guess command-line options.) I think we should show complete `--lsp-server`.

jakecastelli replied:
just to confirm, do we wanna remove `--lsp` from shell completion?

strager reviewed (approved):
I tested with Bash. Mostly looks good.

jakecastelli commented:
thanks strager I am about to rebase, address the comments and test again, will let you know when its ready to land.

jakecastelli commented:
I accidentally broke `powershell` completion, but thanks to `nix`, I am able to test it and fix the bug (I added an extra comma on the last line, kind like `json` and `pwershell` freaked out).

tested on:
- [x] bash
- [x] zsh
- [x] fish
- [x] powershell

Have addressed the comments for `--lsp` and `--path-for-config-search` that does not show autocomplete for file / folders.

Ready to land :shipit: @strager

strager commented on completions/quick-lint-js.ps1:
```diff
@@ -44,7 +44,24 @@
+        @{
+            CompletionText = '--language'
+            ToolTip = "Interpret input files which are given later in the command line as if they were written in languageid`n`ndefault: infer the languageid from the file’s extension`njavascript: the latest ECMAScript standard with proposed features`njavascript-jsx: like javascript but with JSX (React) extensions`nexperimental-typescript: the latest TypeScript version. (EXPERIMENTAL. Subject to  ...
+        },
+        @{
+            CompletionText = '--path-for-config-search'
+            ToolTip = "For the input file or --stdin, use path as the file's path"
+        },
```
`--path-for-config-search` is listed twice. I think this is an accident.
```suggestion
```

jakecastelli commented:
shall we land this? @strager
